src/script/common/c_packer.cpp

```diff
@@ -0,0 +1,566 @@
+			return r;
+		}
+		case LUA_TNUMBER: {
+			auto r = emplace(pv, LUA_TNUMBER);
+			r->ndata = lua_tonumber(L, idx);
+			return r;
+		}
+		case LUA_TSTRING: {
+			auto r = emplace(pv, LUA_TSTRING);
+			size_t len;
+			const char *str = lua_tolstring(L, idx, &len);
+			assert(str);
+			r->sdata.assign(str, len);
+			return r;
+		}
+		case LUA_TTABLE:
+			break; // execution continues
+		case LUA_TFUNCTION: {
+			auto r = emplace(pv, LUA_TFUNCTION);
+			call_string_dump(L, idx);
+			size_t len;
+			const char *str = lua_tolstring(L, -1, &len);
+			assert(str);
+			r->sdata.assign(str, len);
+			lua_pop(L, 1);
+			return r;
+		}
+		case LUA_TUSERDATA: {
+			PackerTuple ser;
+			if (!find_packer(L, idx, ser))
+				throw LuaError("Cannot serialize userdata object");
+			pv.contains_userdata = true;
+			auto r = emplace(pv, LUA_TUSERDATA);
+			r->sdata = ser.first;
+			r->ptrdata = ser.second.fin(L, idx);
+			return r;
+		}
+		default: {
+			std::string err = "Cannot serialize type ";
+			err += lua_typename(L, lua_type(L, idx));
+			throw LuaError(err);
+		}
+	}
+
+	// LUA_TTABLE
+	lua_checkstack(L, 5);
+
+	auto rtable = emplace(pv, LUA_TTABLE);
+	const int vi_table = vidx++;
+
+	lua_pushnil(L);
+	while (lua_next(L, idx) != 0) {
+		// key at -2, value at -1
+		const int ktype = lua_type(L, -2), vtype = lua_type(L, -1);
+		if (ktype == LUA_TNUMBER)
+			rtable->uidata1++; // narr
+		else
+			rtable->uidata2++; // nrec
+
+		// check if we can use a shortcut
+		if (can_set_into(ktype, vtype) && suitable_key(L, -2)) {
+			// push only the value
+			auto rval = pack_inner(L, absidx(L, -1), vidx, pv);
+			rval->pop = vtype != LUA_TTABLE;
+			// and where to put it:
+			rval->set_into = vi_table;
+			if (ktype == LUA_TSTRING)
+				rval->sdata = lua_tostring(L, -2);
+			else
+				rval->sidata1 = lua_tointeger(L, -2);
+			// pop tables after the fact
+			if (!rval->pop) {
+				auto ri1 = emplace(pv, INSTR_POP);
+				ri1->sidata1 = vidx;
+			}
+		} else {
+			// push the key and value
+			auto rkey = pack_inner(L, absidx(L, -2), vidx, pv);
+			vidx++;
+			auto rval = pack_inner(L, absidx(L, -1), vidx, pv);
+			vidx++;
+			// push an instruction to set them
+			auto ri1 = emplace(pv, INSTR_SETTABLE);
+			ri1->set_into = vi_table;
+			ri1->sidata1 = vidx - 2;
+			ri1->sidata2 = vidx - 1;
+			ri1->pop = true;
+			vidx -= 2;
+		}
+
+		lua_pop(L, 1);
+	}
+
+	assert(vidx == vi_table + 1);
+	return rtable;
+}
+
+PackedValue *script_pack(lua_State *L, int idx)
+{
+	if (idx < 0)
+		idx = absidx(L, idx);
+
+	std::unordered_set<const void*> seen;
+	pack_validate(L, idx, seen);
+	assert(seen.size() == 0);
+
+	// Actual serialization
+	PackedValue pv;
+	pack_inner(L, idx, 1, pv);
+
+	return new PackedValue(std::move(pv));
+}
+
+/**** Unpacking implementation ****/
+
+void script_unpack(lua_State *L, PackedValue *pv)
+{
+	const int top = lua_gettop(L);
+	int ctr = 0;
+
+	for (auto &i : pv->i) {
+		if (!i.pop && (ctr++) >= 5) {
+			lua_checkstack(L, 10);
+			ctr = 0;
```

Comment:
What do these magic numbers mean? Would you please be so nice to add a comment?

Reply:
With every instruction there can be up to 2 values left on the stack, and so the space isn't reserved everytime it's only done every 5 iterations. Will add a comment.